intel-oneapi-mkl: do not set __INTEL_POST_CFLAGS env variable

balay opened the pull request:
```
petsc configure breaks with spack change that now adds in __INTEL_POST_CFLAGS - as it rquires compilers not to print spurious warnings

ref: xsdk@pj02:~/spack$ ./bin/spack spec petsc~fortran~metis~hypre~hdf5 Input spec
--------------------------------
 -   petsc~fortran~hdf5~hypre~metis

Concretized
--------------------------------
[+]  petsc@3.20.1%oneapi@2022.2.0~X~batch~cgns~complex~cuda~debug+double~exodusii~fftw~fortran~giflib~hdf5~hpddm~hwloc~hypre~int64~jpeg~knl~kokkos~libpng~libyaml~memkind~metis~mkl-pardiso~mmg~moab~mpfr+mpi~mumps~openmp~p4est~parmmg~ptscotch~random123~rocm~saws+shared~strumpack~suite-sparse~sycl~tetgen~trilinos~valgrind build_system=generic clanguage=C memalign=none arch=linux-ubuntu20.04-skylake
[e]      ^diffutils@3.7%oneapi@2022.2.0 build_system=autotools arch=linux-ubuntu20.04-skylake
[+]      ^gmake@4.4.1%oneapi@2022.2.0~guile build_system=generic arch=linux-ubuntu20.04-skylake
[e]      ^intel-oneapi-mkl@2022.2.0%oneapi@2022.2.0~cluster+envmods~ilp64+shared build_system=generic mpi_family=none threads=none arch=linux-ubuntu20.04-skylake
[e]      ^intel-oneapi-mpi@2021.7.0%oneapi@2022.2.0+envmods~external-libfabric~generic-names~ilp64 build_system=generic arch=linux-ubuntu20.04-skylake
[e]      ^python@3.9.7%oneapi@2022.2.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=0d98e93,4c24573,ebdca64,f2fd060 arch=linux-ubuntu20.04-skylake

__INTEL_POST_CFLAGS=-Wl,-rpath,/opt/intel/oneapi/mkl/2022.2.0/lib/intel64; export __INTEL_POST_CFLAGS __INTEL_POST_FFLAGS=-Wl,-rpath,/opt/intel/oneapi/mkl/2022.2.0/lib/intel64; export __INTEL_POST_FFLAGS

Executing: /opt/intel/oneapi/mpi/2021.7.0/bin/mpiicc -E  -I/tmp/petsc-km5zdzsi/config.setCompilers  /tmp/petsc-km5zdzsi/config.setCompilers/conftest.c Possible ERROR while running preprocessor:exit code 0

stderr:
icx: warning: -Wl,-rpath,/opt/intel/oneapi/mkl/2022.2.0/lib/intel64: 'linker' input unused [-Wunused-command-line-argument]
```

Prior discussion at https://github.com/spack/spack/pull/35737#issuecomment-1796958712

Likely trigger for this change in spack: https://github.com/spack/spack/commit/e5f3ffc04fb7a1fea237226210b0cafc9246d0f2

balay commented:
@eugeneswalker CI is likely to fail here - as it did at https://github.com/spack/spack/pull/40825#issuecomment-1801957922

haampie commented:
Can you check where these env variables are coming from?

Probably from `setup_run_environment` in `intel-oneapi-mkl` or `intel-oneapi-mpi`.

If that is breaking your package then probably those packages should be fixed instead of every dependent.

balay commented:
> Probably from setup_run_environment in intel-oneapi-mkl

yes

balay commented:
And removing `setup_run_environment()` from `intel-oneapi-mkl` gets this build working for me.

eugeneswalker commented:
> @eugeneswalker CI is likely to fail here - as it did at [#40825 (comment)](https://github.com/spack/spack/pull/40825#issuecomment-1801957922)

I'm sorry for the inconvenience here @balay. When I saw your issue report yesterday, I fixed it on our Roberta Cray machine. Seeing it here now, also affecting our Gary Cray machine, I've fixed it there too just now. 

I retried those failed petsc builds you linked to after fixing, and they succeeded.
* https://gitlab.spack.io/spack/spack/-/jobs/9021534 (success on manual retry, post fix)
* https://gitlab.spack.io/spack/spack/-/jobs/9021535 (success on manual retry, post fix)

We shouldn't see this anymore. Sorry for the inconvenience!

balay commented:
And I see that change is from #30981

cc: @skosukhin

balay commented:
> We shouldn't see this anymore

Great! Thanks @eugeneswalker !

haampie commented:
I'm confused, where is this fixed?

balay commented:
> I'm confused, where / how is this fixed? (@eugeneswalker)

that's regarding a different issue with CI failures [with petsc] 

https://github.com/spack/spack/pull/40825#issuecomment-1801957922

balay commented:
> If that is breaking your package then probably those packages should be fixed instead of every dependent.

Alternative: what packages actually need these env variables set - for use with intel-oneapi-mkl. Perhaps they have broken build systems - that need fixing [or selectively set these flags for them]?

skosukhin commented:
Do I understand correctly that the compilation fails because a warning is promoted to an error?

balay commented:
> Do I understand correctly that the compilation fails because a warning is promoted to an error?

In some sense that's what a CPP always does [as stdout is reserved for pre-processed output - and `stderr` is used for any compiler messages]


Notice: 

Executing: /opt/intel/oneapi/mpi/2021.7.0/bin/mpiicc `-E`  ...

`stderr`:

icx: warning: -Wl ....

skosukhin commented:
@balay what I mean is that warnings are usually emitted to `stderr` but do not alter the zero exit code of a program (do not halt the preprocessing in this case). The warnings can be made errors though. For example, with the `-Werror` flag. There might be other ways to trigger that. So my question was whether that's the case here. Maybe the flag is in the `compilers.yaml` or maybe the build system appends it at some point?

balay commented:
petsc configure attempts to determine compiler flags that CPP accepts - or rejects - so it has to rely on these messages to successfully execute such tests.

skosukhin commented:
@balay I see, I assume that the warning is intentionally enabled (taking [this sentence](https://github.com/spack/spack/blob/afc693645a2b523ec45bc781bbfb5b29a3e7c680/var/spack/repos/builtin/packages/intel-oneapi-compilers/package.py#L266-L267) into account). Thanks for the clarification.

@haampie do you know if it's possible to have [these environment modifications](https://github.com/spack/spack/blob/afc693645a2b523ec45bc781bbfb5b29a3e7c680/var/spack/repos/builtin/packages/intel-oneapi-mkl/package.py#L155-L165) in the generated environment modules but not in the Spack build environment? The variables are useful outside of Spack but are at best redundant within it. In the worst case, I don't mind dropping them (they were introduced by me in #30981).

balay commented:
>  I assume that the warning is intentionally enabled

Not within petsc configure. That appears to be the default compiler behavior. For ex: with regular - non-spack env

```console

xsdk@pj02:~/junk$ I_MPI_CC=icx /opt/intel/oneapi/mpi/2021.7.0/bin/mpiicc -E a.c > /dev/null
xsdk@pj02:~/junk$ __INTEL_POST_CFLAGS=-Wl,-rpath,/opt/intel/oneapi/mkl/2022.2.0/lib/intel64 I_MPI_CC=icx /opt/intel/oneapi/mpi/2021.7.0/bin/mpiicc -E a.c > /dev/null
icx: warning: -Wl,-rpath,/opt/intel/oneapi/mkl/2022.2.0/lib/intel64: 'linker' input unused [-Wunused-command-line-argument]
xsdk@pj02:~/junk$ 
```

skosukhin commented:
@haampie it's in the MKL package, not the compiler one.

haampie commented:
Hm. Could you move the variable to `modules.yaml` config for site specific use? Or is that annoying too?

skosukhin commented:
Mmm :) Ok. I don't mind if someone removes that stuff from the MKL package.

haampie commented:
could you turn this PR into updating the MKL package @balay?

balay committed:
intel-oneapi-mkl: do not set __INTEL_POST_CFLAGS env variable

This triggers warnings from icx compiler - that breaks petsc configure

$ I_MPI_CC=icx /opt/intel/oneapi/mpi/2021.7.0/bin/mpiicc -E a.c > /dev/null
$ __INTEL_POST_CFLAGS=-Wl,-rpath,/opt/intel/oneapi/mkl/2022.2.0/lib/intel64 I_MPI_CC=icx /opt/intel/oneapi/mpi/2021.7.0/bin/mpiicc -E a.c > /dev/null
icx: warning: -Wl,-rpath,/opt/intel/oneapi/mkl/2022.2.0/lib/intel64: 'linker' input unused [-Wunused-command-line-argument]
$

balay commented:
> could you turn this PR into updating the MKL package @balay?

Ok removed my prior change to `petsc` - and added in the change to `intel-oneapi-mkl`
